flink-table/flink-table-storage/src/main/java/org/apache/flink/table/storage/filestore/combine/SortMergeReader.java

```diff
@@ -0,0 +1,158 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.storage.filestore.combine;
+
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.table.storage.filestore.KeyValue;
+import org.apache.flink.table.storage.filestore.utils.RecordIterator;
+import org.apache.flink.table.storage.filestore.utils.RecordReader;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Objects;
+import java.util.PriorityQueue;
+
+/**
+ * This reader is to read a list of {@link RecordReader}, which is already sorted by key and
+ * sequence number, and perform a sort merge algorithm. Note that values are not combined at this
+ * phase.
+ */
+public class SortMergeReader implements RecordReader<KeyValue> {
+    private final List<RecordReader<KeyValue>> sortedReaders;
+    private final Comparator<RowData> userKeyComparator;
+    private PriorityQueue<Element> minHeap;
+    private Element current;
+
+    public SortMergeReader(
+            List<RecordReader<KeyValue>> sortedReaders, Comparator<RowData> userKeyComparator) {
+        this.sortedReaders = sortedReaders;
+        this.userKeyComparator = userKeyComparator;
+    }
+
+    @Nullable
+    @Override
+    public RecordIterator<KeyValue> readBatch() throws IOException {
+        if (sortedReaders.isEmpty()) {
+            return null;
+        }
+        if (Objects.isNull(minHeap)) {
+            heapify();
+        } else {
+            supply();
+        }
+        if (minHeap.size() > 0) {
+            return new SortMergeIterator();
+        } else {
+            return null;
+        }
+    }
+
+    @Override
+    public void close() throws IOException {
+        for (RecordReader<KeyValue> reader : sortedReaders) {
+            reader.close();
+        }
+    }
+
+    private void heapify() throws IOException {
+        minHeap =
+                new PriorityQueue<>(
+                        (e1, e2) -> {
+                            int result =
+                                    userKeyComparator.compare(
+                                            e1.iterator.current().key(),
+                                            e2.iterator.current().key());
+                            if (result != 0) {
+                                return result;
+                            }
+                            return Long.compare(
+                                    e1.iterator.current().sequenceNumber(),
+                                    e2.iterator.current().sequenceNumber());
+                        });
+        for (int i = 0; i < sortedReaders.size(); i++) {
+            RecordIterator<KeyValue> iterator = sortedReaders.get(i).readBatch();
+            if (Objects.nonNull(iterator) && iterator.advanceNext()) {
```

Comment:
Objects.nonNull and isNull is used used as a {@link java.util.function.Predicate}, {@code filter(Objects::isNull)}. (In method document)